### ISSUE TITLE
안녕하세요. 최근에 딥러닝을 공부하고있는 직장인입니다. 로스함수 관련 궁금한게 있어 문의드립니다. 

### ISSUE THREAD

**Issue by gellston** [swyoon/normalized-autoencoders#4]
안녕하십니까

성남에서 검사관련 회사에서 일하고있는 직장인입니다. 
딥러닝에 관심이 많아 최근에 딥러닝을 이용하여 anomality detection모델을 개인적으로 공부하면서 이것저것 찾다가
tensorflow kr 페이스북에서 해당 사이트 링크를 타고 들어와서 논문을 보게되었습니다.

현재 제가 anomality 검출 모델을 공부하는 중에 시도해본 모델은 일반적인 컨볼루션 오토인코더를 이용해서 
입력 대비 출력간의 편차를 이용하여 검출해보고자 했었습니다. 하지만 논문에서도 말씀하셨다시피 이물, 혹은 outliear 데이터에 대해 defeat 부분이 필터링되지 않고 출력으로 나오는 경우가 있어 고민중에 있었습니다. 

작성하신 논문에서는 이부분을 큰 수정없이 획기적으로 개선한것 같아보여(유투브 동영상을 시청했었습니다.)
이를 참고해서 개인적으로 공부를 해보고자 하는데 

![image](https://user-images.githubusercontent.com/42884297/133020204-72a20cb8-3d51-43fe-8855-b74e7accc479.png)

동영상에서도 말씀하신 minimize reconstruction error부분은 일반적인 mean square error로 보이고 
Enforcing normalization 텀쪽의 오메가 쪽 부분을 보면 exponential에 -E0(mean square error)를 넣고 모든 픽셀에 대해서 sum을 한 이후 (integral) 이를 log에 넣고 minimize reconstruction error항과 더해서 최종 토탈 loss를 구하신것으로 이해했습니다.

![image](https://user-images.githubusercontent.com/42884297/133020169-9fa7392f-cd7f-42af-9a51-69bfbfd0e1ec.png)

하지만 실제 keras로 구현하여 학습해보았을때 loss가 0로 나오고 학습이 되지않아 제가 이해한 것이 맞는지 
여쭤보기 위해서 이렇게 질문드립니다. 토치 버전의 trainer 쪽 코드도 참고하였으나 동영상에서 언급하신 log(오메가)항이 없고 
다르게 코드가 구현되어있는것 같아 혼란스럽습니다. 제가 만약 틀리게 이해를 한것으라면 혹시 팁을 주실수있으실가요? 

![image](https://user-images.githubusercontent.com/42884297/133021305-40675635-0159-4102-ab0e-68a2ee040d06.png)

**Comment by swyoon** [comment]
@gellston 님, 안녕하세요? 우선 본 연구에 관심가져주셔서 대단히 감사드립니다.
해당 영상은 홍보 목적으로 만들어진지라, 아무래도 구현에 대한 자세한 해설이 없어서 영상과 코드만으로는 파악이 어려우실 것 같습니다ㅠㅠ
일단은 간단한게 답변드릴 수 있는 부분들을 먼저 답드리겠습니다.

1. 오메가의 적분은 모든 픽셀에 대한 적분이 아니라 가능한 모든 입력값에 대한 적분입니다. 즉, 이미지의 경우 가능한 모든 pixel값들에 대한 적분입니다.
2. 당연하게도, 해당 적분은 직접 계산이 불가능합니다. 따라서 코드에도 log Omega는 등장하지 않습니다ㅠㅠ
3. 본 알고리즘에서 log Omega는 "암묵적인 방식으로" 계산됩니다. 어떤 암묵적인 방식이냐면, 모델로부터 negative sample을 만들어내어서, 이미지들에 대한 reconstruction error을 증가시키는 방법입니다.

해당하는 전체 과정에 대한 상세한 설명은 논문에 나와있긴 하지만, Energy-based Model이나 Markov Chain Monte Carlo (MCMC) 기법에 대한 약간의 배경지식이 있으셔야 이해가 쉬울 듯 합니다.
MCMC를 사용하지 않는, 더 간단한 학습 방법들에 대해서는 제가 현재 연구를 진행 중에 있습니다.

추가적으로 더 궁금하신 부분이 있으시면, 남겨주시거나 연락주시면 제가 최대한 답변해드릴 수 있도록 하겠습니다.

**Comment by gellston** [comment]
친절한 설명 감사드립니다. 알려주신 해당 키워드로 한번 공부해보도록하겠습니다. 감사합니다.
